pluginbase.py

```diff
@@ -1,37 +1,28 @@
+import typing
+
 from abc import ABC
 from abc import abstractmethod
 from typing import Optional
 
-import perf
-
-from task import Task
-from testConfig import TestConfig
 from tftbase import PluginOutput
 from tftbase import PluginResult
 from tftbase import TestMetadata
 
 
-class PluginTask(Task):
-    @property
-    @abstractmethod
-    def plugin(self) -> "Plugin":
-        pass
-
-
 class Plugin(ABC):
     PLUGIN_NAME = ""
 
     @abstractmethod
     def enable(
         self,
         *,
-        tc: TestConfig,
+        tc: "TestConfig",
```

Comment:
This just seems a very roundabout way to handle dependency issues.... Could we discuss any cleaner alternatives?

Reply:
If we had a ABC for perf and others, would it solve some of the circular dependency issues? Also if we move some of the classes to their own files?

WDYT?

Reply:
The patch solves a loop of the "import". That was an actual problem.

Beyond that, we still have a loop, for the typing hints and which types are referred to.

Maybe you could do something with moving classes around, and introducing new ABC classes (that have fewer dependencies themselves). I doubt it however, because implementations of the virtual `pluginbase.Plugin.enable()` want an actual `perf.PerfServer` type and not `perf_server: PerfServerABC`.

And if you think, "but I don't understand why there is `if typing.TYPE_CHECKING: from task import PerfServer`", then just try to remove it. Then you will note that mypy will complain about an issue, and you will make it obvious why it's there. Maybe then you can come up with solutions, that move code around, introduce an ABC and involve fewer/no `if typing.TYPE_CHECKING:`. That might be interesting (if possible), but there is no real problem solved by that. The current solution is IMO fine, it has no problematic import loop and typing hints fully work.

See also:
- https://mypy.readthedocs.io/en/stable/runtime_troubles.html#import-cycles
- https://www.reddit.com/r/learnpython/comments/zzgkxj/comment/j2bmhlj/

Reply:
ah, `pluginbase.PluginTask` had to move to `task.PluginTask`, because it uses a `Task` for `class PluginTask(Task)`. it's not merely for typing hints, it's in actual use.

And, as we need to break the import loop between `pluginbase`->`task`->`testConfig`->`pluginbase`, something needs to move.

Since it makes sense that `testConfig` uses `pluginbase.Plugin` and that `task` uses `testConfig.TestConfig`, the chosen solution is that `pluginbase` does not import `task`. Hence, `PluginTask` cannot be in `pluginbase` and had to move (as it requires `task.Task` -- not only for typing hints).

Reply:
Let's merge your PR in for now. I understand what you are trying to say. Maybe there is more work to decouple the dependencies and objects. But it is messy right now, not due to your changes but in general how we structured the inheritance.

Reply:
@wizhaoredhat thank you.

> But it is messy right now, not due to your changes but in general how we structured the inheritance.

This particular place is IMO *not* messy and there is nothing to improve (other places are IMO messy). Of course, if you come up with a patch that proves me wrong, then great...